### PR TITLE
Optionally (by default) download & build stripped arrow as part of rerun_sdk 

### LINKED ISSUE #4028: Investigate whether we can build arrow-cpp from source in C++
That would put all those complicated stdlib ABI questions to rest and make the C++ onboarding process that much smoother.

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes  #4028

Confirmed to be working on Linux, Mac and Windows. On windows also made sure this works with debug runtime and experimented a bit with Visual Studio setups. Also checked Windows against the external sample, see https://github.com/rerun-io/cpp-example-opencv-eigen/pull/8

Made improvements to our CMakeLists files overall. Notably we no longer need `DOWNLOAD_EXTRACT_TIMESTAMP ON` to suppress warnings - idea from https://github.com/rerun-io/cpp-example-opencv-eigen/pull/6, setting a range for the cmake version fixes it

Drawbacks:
* LOTS of cmake log spam during build
* a bit longer builds on the first run, seems to be a bit more pronounced on Windows (no hard numbers available yet)

Advantages:
* no longer needed to install libarrow
* works on Windows debug builds (note that other libraries like OpenCV might still have this issue)
* stable against all c++ library/compiler changes, no fear of weird runtime errors because of stdlib mismatches
* more lightweight artifacts - on Windows one would have needed to ship a large number of dlls, now none

It's still worth considering if we should ship prebuilt debug & release arrow dlls instead.

TODO: We need to document this new behavior and how to turn it off in the respective doc pages. By default the user no longer needs to install libArrow with this.

Important CMake options that need documenting (can partially copy& paste this!):
* `RERUN_ARROW_LINK_SHARED`
    * exited before but now it's default OFF (EDIT:) on windows. This makes it easier to relocate windows executable (don't need to copy Arrow.dll around!). EDIT: On mac this made the address sanitizer crazy for me and I reckon it's an unnecessary slowdown
* `RERUN_ARROW_EXTERNAL_PROJECT` (new!)
    * Default on
    * If enabled we download Arrow 10.0.1 and build it in a as small as possible fashion
    * If disabled we run `find_package` == old behavior
* `RERUN_C_LIB`
   * where to find the static c library that rerun_sdk links against
   * in repo defaults to cargo built
   * outside repo defaults to /lib/ folder + platform artifact name

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/4051) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4051)
- [Docs preview](https://rerun.io/preview/e38502cfb40b406f9d22c5bb8404ea1183ea2cdd/docs) 
- [Examples preview](https://rerun.io/preview/e38502cfb40b406f9d22c5bb8404ea1183ea2cdd/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: rerun_cpp/CMakeLists.txt [hunk 1]
```diff
@@ -97,33 +90,133 @@ if("${RERUN_C_LIB}" STREQUAL "")
...
+        -DARROW_BUILD_SHARED=${ARROW_BUILD_SHARED}
+        -DARROW_BUILD_STATIC=${ARROW_BUILD_STATIC}
+        -DCMAKE_INSTALL_PREFIX=${ARROW_DOWNLOAD_PATH}
+        -DARROW_USE_ASAN=OFF
+        -DARROW_USE_TSAN=OFF
+        -DARROW_USE_UBSAN=OFF
+        -Dxsimd_SOURCE=BUNDLED
+        -DBOOST_SOURCE=BUNDLED
+        -DARROW_BOOST_USE_SHARED=OFF
+        -DARROW_JEMALLOC=OFF
```
Comment (Wumpf) [inline rerun_cpp/CMakeLists.txt:173]: move this up a little bit, looks nicer together with the other simple off stuff

File: rerun_cpp/CMakeLists.txt [hunk 2]
```diff
@@ -97,33 +90,133 @@ if("${RERUN_C_LIB}" STREQUAL "")
...
+
+target_link_libraries(rerun_sdk PRIVATE rerun_c)
 
 # -----------------------------------------------------------------------------
 # Arrow:
-option(RERUN_ARROW_LINK_SHARED "Link to the Arrow shared library" ON)
+# Using the static arrow library by default.
+# This makes the setup a lot easier on Windows where we otherwise need to put Arrow.dll either in path or copy it with the exectuable.
+option(RERUN_ARROW_LINK_SHARED "Link to the Arrow shared library" OFF)
+option(RERUN_ARROW_EXTERNAL_PROJECT "If enabled, arrow will be added as an external project and built with the minimal set required by the Rerun C++ SDK" ON)
```
Comment (jleibs) [inline rerun_cpp/CMakeLists.txt:113]: Nit: although `ExternalProject` is the name of the utility, `RERUN_ARROW_EXTERNAL_PROJECT` sounds like the exact opposite of "Build arrow as part of the rerun project"

Something like `RERUN_BUILD_ARROW` makes more sense to me.

Comment (Wumpf) [reply rerun_cpp/CMakeLists.txt:113]: or RERUN_DOWNLOAD_AND_BUILD_ARROW maybe

File: rerun_cpp/CMakeLists.txt [hunk 3]
```diff
@@ -97,33 +90,133 @@ if("${RERUN_C_LIB}" STREQUAL "")
...
+    if(RERUN_ARROW_LINK_SHARED)
+        message(STATUS "Arrow SO version: ${ARROW_FULL_SO_VERSION}")
+        add_library(RerunArrowTarget ALIAS Arrow::arrow_shared)
+    else()
+        message(STATUS "Arrow version: ${ARROW_VERSION}")
+        add_library(RerunArrowTarget ALIAS Arrow::arrow_static)
     endif()
 
-    target_link_libraries(rerun_sdk PRIVATE Arrow::arrow_static)
+    set(RERUN_ARROW_TARGET Arrow::arrow_shared)
```
Comment (jleibs) [inline rerun_cpp/CMakeLists.txt:212]: Shouldn't this me inside the `if` with
```
set(RERUN_ARROW_TARGET Arrow::arrow_static)
```
in the `else`?

Comment (jleibs) [reply rerun_cpp/CMakeLists.txt:212]: Or maybe this isn't used at all?

**Review by jleibs — APPROVED** [review]
Tested this on linux -- works for me and looks great. Much as I love pixi this really lowers the bar for ease of integration.

File: rerun_cpp/CMakeLists.txt (hunk 3 above)
Comment (Wumpf) [reply rerun_cpp/CMakeLists.txt:217]: oh. yeah that's a broken left over

**Pull request merged**